[[ AndroidAudioRecorder ]] Add android audio recorder library

livecodeali opened the pull request:
This patch adds an LCB library which gives access to the Android
MediaRecorder API, allowing audio recording on Android. It also
includes a sample stack demonstrating the feature.

livecodeali committed:
[[ AndroidAudioRecorder ]] Add android audio recorder library
This patch adds an LCB library which gives access to the Android
MediaRecorder API, allowing audio recording on Android. It also
includes a sample stack demonstrating the feature.

runrevmark commented on extensions/libraries/androidaudiorecorder/androidaudiorecorder.lcb:
```diff
@@ -0,0 +1,329 @@
+    end if
+    
+    variable tFile as JString
+    put GetAbsolutePathToPublicDocument(pFileName) into tFile
```
Is there a reason we restrict this to external storage? What if the incoming path is absolute already?

livecodeali replied:
Heh, good point... actually external storage path should probably be in an android utility library

runrevmark replied:
As long as the external storage path is accessible via specialFolderPath() and the recorder functions report a suitable error if the file can't be created; we can probably defer any such processing to user script.

livecodeali replied:
I don't think the external storage path is accessible via specialFolderPath...

runrevmark replied:
I think it should be:
```
	static final String s_external_prefix = "external ";
	public String getSpecialFolderPath(String p_name)
	{
		boolean t_external = false;
		if (p_name.startsWith(s_external_prefix))
		{
			t_external = true;
			p_name = p_name.substring(s_external_prefix.length());
		}

		try
		{
			if (p_name.equalsIgnoreCase("documents"))
			{
				if (t_external)
					return getContext().getExternalFilesDir(null).getAbsolutePath();
				else
					return getContext().getFilesDir().getAbsolutePath();
			}
			else if (p_name.equalsIgnoreCase("temporary") || p_name.equalsIgnoreCase("cache"))
			{
				if (t_external)
					return getContext().getExternalCacheDir().getAbsolutePath();
				else
					return getContext().getCacheDir().getAbsolutePath();
			}
		}
		catch (Exception e)
		{
			return "";
		}

		return "";
	}
```
Via 'external documents' - maybe? I think that's the same call you are using in LCB.

livecodeali replied:
Ah! It is not documented!

runrevmark commented:
My only query is about the 'filename' which is passed - we are automatically prepending the external storage location. I wonder if we should just make the filename be processed as it would normally for engine commands/functions which take filenames?

livecodeali committed:
[[ AndroidAudioRecorder ]] Don't automatically write to external storage
- Expect an absolute path to the record filename to be specified.
- Remove automatic addition of 'WRITE_EXTERNAL_STORAGE' permission.
- Modify the sample stack to write recording to external storage path

runrevmark commented:
@livecode-vulcan review ok be9d708

livecode-vulcan commented:
💙 review by @runrevmark ok be9d708360d53ae0825ce256d37f2f9f97ac4ad6

livecode-vulcan commented:
😎 test success be9d708360d53ae0825ce256d37f2f9f97ac4ad6

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/767)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/783)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/757)
* try-community-universal-ios-iphoneos10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos10.3/builds/701)
* try-community-universal-ios-iphonesimulator10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator10.3/builds/704)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/761)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/783)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/766)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/783)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/765)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/776)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/797)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/761)
